### PR TITLE
Validate data write size limit in ZkClient

### LINKED ISSUE #1071: Znode write size limit in zkclient is not consistent with the one in znrecord serializers
There is configurable write size limit for znrecord serializer. But there is also a data size limit check in raw zkclient:
```
private void checkDataSizeLimit(byte[] data) {
    if (data != null && data.length > ZNRecord.SIZE_LIMIT) {
```
If configurable write size limit for znrecord serializer is different from the default size 1 MB, especially it is set to larger than 1 MB, writing won't work because it would be blocked by this default size check in zkclient.

We need to fix it. A good idea is to make both consistently same. But the system property `zk.serializer.znrecord.write.size.limit.bytes` is for a **ZNRecord serializer**, not sure if it is fine to put it at ZkClient level, as it will apply to all other serializers, as well. We may discuss this.

### PR DESCRIPTION
**Pull request by pkuwm**
### Issues

- [x] My PR addresses the following Helix issues and references them in the PR description:

Fixes #1071 

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

If configurable write size limit for znrecord serializer is different from the default size 1 MB, especially it is set to larger than 1 MB, writing won't work because it would be blocked by this default size check in zkclient.

This PR also reads both `jute.maxbuffer` config and znrecord serializer write size limit, and validates the correctness of data size limit.

### Tests

- [x] The following tests are written for this issue:

- testDataSizeGreaterThanLimit
- testDataSizeLessThanLimit
- testInvalidWriteSizeLimitConfig

- [x] The following is the result of the "mvn test" command on the appropriate module:

In zookeeper-api module.
```
[INFO] Tests run: 50, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 64.326 s - in TestSuite
[INFO]
[INFO] Results:
[INFO]
[INFO] Tests run: 50, Failures: 0, Errors: 0, Skipped: 0
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  01:08 min
[INFO] Finished at: 2020-10-22T17:17:17-07:00
[INFO] ------------------------------------------------------------------------
```

### Commits

- [x] My commits all reference appropriate Apache Helix GitHub issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation (Optional)

- [x] In case of new functionality, my PR adds documentation in the following wiki page:

(Link the GitHub wiki you added)

### Code Quality

- [x] My diff has been formatted using helix-style.xml 
(helix-style-intellij.xml if IntelliJ IDE is used)

### CODE REVIEW

File: helix-core/src/test/java/org/apache/helix/manager/zk/TestRawZkClient.java
Comment (narendly) [inline helix-core/src/test/java/org/apache/helix/manager/zk/TestRawZkClient.java:824]: Thanks for adding this test! One favor to ask - we are actually in the process of moving this class to zookeeper-api (https://github.com/apache/helix/pull/1070).

Do you think we could hold off on this change until that is committed first to avoid merge conflicts?

File: zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java
Comment (narendly) [inline zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java:1858]: I agree that this is the right thing to do. Thanks for catching this and fixing this.

One concern I have is with the change of the exception type. I am not sure if changing it to ZkMarshallingError is the right thing to do because

1. ZkMarshallingError is usually thrown in ZkSerializer implementations.
2. Changing Exception types may cause backward-compatibility issues.

What do you think?

Comment (pkuwm) [reply zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java:1858]: These questions are valid and natural ones which I also thought about.
1. The reason I change it is I would like to make it consistent with serializer. In helix zkclient, the data size check is right after serializing. `ZkMarshallingError` could also indicate the serialized data still exceeds the limit. So I feel `ZkMarshallingError` is more accurate than `ZkClientException`.
2. Actually before version 1.0.0, the exception type is `HelixException`. So changing to `ZkClientException` already breaks backward-compatibility if it breaks. Considering recent 1.0.0 zookeeper-api module users(helix-core 1.0.0 is deprecated), if a user explicitly uses this raw zkclient and catches this runtime exception, that may break. But I am not sure if that really exists in 1.0.0 usage, considering the APIs don't explicitly signature `ZkClientException`.

Though having these thoughts, I don't have a strong preference: we could keep the original exception type. If we would like to change the type, may be it is a good idea to do it in 1.0.1 to avoid future backward compatibility issue.

Comment (narendly) [reply zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java:1858]: To give you more context, ZkMarshallingError is something that was inherited from 101tec.zkclient library and explicitly thrown for ZkSerializer implementations. Since this is our ZkClient API-level check outside the serializer, I think ZkClientException. 

Also, we've decided to convert all HelixExceptions to ZkClientException, so I think ZkClientException would be more appropriate.

Comment (dasahcc) [reply zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java:1858]: +1. Let's not change the exception type of this.

Comment (pkuwm) [reply zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java]: Thanks for the input. Resolved.

**Comment by jiajunwang** [comment]
I'm not convinced that this is a good idea.

For the non-Helix user, ZNRecord is not meaningful. How do you plan to document this config and let them configure ZK_SERIALIZER_**ZNRECORD**_WRITE_SIZE_LIMIT_BYTES for the general ZK usage?
Secondly, we shall not block the write in the ZkClient if the write size is smaller than the jute.maxbuffer configuration. That is not very meaningful. On the other hand, if we have a large ZK_SERIALIZER_ZNRECORD_WRITE_SIZE_LIMIT_BYTES configured, but the jute.maxbuffer is small, what we shall do? The write will still fail.

My suggestion,
- The raw ZkClient shall check size based on jute.maxbuffer, and maybe we leave some buffer like the default limit is current doing.
- If jute.maxbuffer is not configured, then we apply the default value according to the ZK document.

**Comment by pkuwm** [comment]
> I'm not convinced that this is a good idea.
> 
> For the non-Helix user, ZNRecord is not meaningful. How do you plan to document this config and let them configure ZK_SERIALIZER_**ZNRECORD**_WRITE_SIZE_LIMIT_BYTES for the general ZK usage?
> Secondly, we shall not block the write in the ZkClient if the write size is smaller than the jute.maxbuffer configuration. That is not very meaningful. On the other hand, if we have a large ZK_SERIALIZER_ZNRECORD_WRITE_SIZE_LIMIT_BYTES configured, but the jute.maxbuffer is small, what we shall do? The write will still fail.
> 
> My suggestion,
> 
> * The raw ZkClient shall check size based on jute.maxbuffer, and maybe we leave some buffer like the default limit is current doing.
> * If jute.maxbuffer is not configured, then we apply the default value according to the ZK document.

@jiajunwang Thanks for the comments. I have the same concern about using the same config at zkclient level. This is why I would like to have it discussed at #1071 and confirmed: 
> A good idea is to make both consistently same. But the system property zk.serializer.znrecord.write.size.limit.bytes is for a ZNRecord serializer, not sure if it is fine to put it at ZkClient level, as it will apply to all other serializers, as well. We may discuss this.

I have one thought if we use `jute.maxbuff` value here. If `jute.maxbuff` is set, considering client and server's values are the same. zkclient might be able to write but fail to read, due to the extra fields in ZK server. This is why we want to leave some buffer. 

But it may be difficult to determine the buffer size. Say we leave 5 KB as buffer, the value would be maxbuffer - 5 KB.  But the value of ZK_SERIALIZER_**ZNRECORD**_WRITE_SIZE_LIMIT_BYTES might fall in the range of [maxbuffer, maxbuffer - 5 KB]. Even serializer check passes but it could be blocked by this check maxbuffer - 5 KB.

If we are using jute.maxbuffer for this check, we have to document to make it clear  ZK_SERIALIZER_ZNRECORD_WRITE_SIZE_LIMIT_BYTES should be less or equal to maxbuffer - 5KB.

If possible, I would prefer to just have one config `zkclient.znode.write.size.limit` at ZkClient level, instead of znrecord serializer. So it satisfies all cases. Let's discuss further.

**Comment by jiajunwang** [comment]
1. jute.maxbuffer is a Zookeeper lib configuration, I don't think we need to add additional doc about it.
2. Regarding how to determine the buffer, I think we can just do what we are doing now. Keeping the behavior the same.
3. To avoid confusion, we can do the configuration check when the ZkClient is constructed. Just fail the constructing if the ZK_SERIALIZER_ZNRECORD_WRITE_SIZE_LIMIT_BYTES is configured larger than the ZkClient write limitation. With a clear Exception, users will understand what to do.

**Comment by pkuwm** [comment]
> * jute.maxbuffer is a Zookeeper lib configuration, I don't think we need to add additional doc about it.
> * Regarding how to determine the buffer, I think we can just do what we are doing now. Keeping the behavior the same.
> * To avoid confusion, we can do the configuration check when the ZkClient is constructed. Just fail the constructing if the ZK_SERIALIZER_ZNRECORD_WRITE_SIZE_LIMIT_BYTES is configured larger than the ZkClient write limitation. With a clear Exception, users will understand what to do.

@jiajunwang Thanks for the input. Size margin would be using the current margin: default 1 MB - ZNRecord size limit (1000 KB) = 24 KB.  I've updated the PR.

File: zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java
Comment (jiajunwang) [inline zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java:89]: I'm not sure if we can directly use the value ClientCnxn.packetLen. It is public and static.

File: zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java
Comment (jiajunwang) [inline zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java:2252]: This is awkward. How about just create a constant for margin and deprecate the SIZE_LIMIT?

File: zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java
Comment (jiajunwang) [inline zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java:2245]: With compression on, the serializer may always return a much smaller package than the input size.
So do we have to fail it here?

It is debatable.

File: zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java
Comment (pkuwm) [reply zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java:2252]: ZNRecordSerializer is also using the SIZE_LIMIT to determine the actual size limit. Maybe we still need it. Otherwise, we still have `sizeLimit = JUTE_MAXBUFFER - MARGIN`. I understand this calculation is awkward, but just it is just a local variable. Or we could add one more constant for margin so we don't have to calculate it again here.

File: zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java
Comment (pkuwm) [reply zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java:89]: Answer is no. That one is different: it's for the response packet size and it's 4 MB. We are looking at server side jute.maxbuffer that is by default 1 MB in `BinaryInputArchive`. It does no harm to define our own private constant and the default value 1 MB is known.

File: zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java
Comment (pkuwm) [reply zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java:2245]: In this case, we could add one more check for compression on/off. If compression is turned on, we allow serializerSize > zkClientSize. Agreed?

File: zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java
Comment (jiajunwang) [reply zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java:2252]: The main concern is that you introduced some internal logic that is not configurable or commented directly where the variables are defined. Who will be able to tell the margin is derived by the default max buffer size and the default size limit?
First, this logic is not reasonable, why the margin is based on the default configs? Secondly, it introduces more chaos to our code.

Comment (jiajunwang) [reply zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java:2252]: Just add a MARGIN value please.

File: zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java
Comment (jiajunwang) [reply zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java:89]: What if ZK changes the default max buffer in some new versions? Are we going to change this number accordingly? So my suggestion is that,
1. We fine the default value in Zookeeper code.
2. if 1 is not possible, then we can just use the current SIZE_LIMIT for the default value.

File: zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java
Comment (jiajunwang) [inline zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java:2256]: So according to what I comment before, the logic here can be,
if (value of ZkSystemPropertyKeys.JUTE_MAXBUFFER exits) {
  return value of ZkSystemPropertyKeys.JUTE_MAXBUFFER - MARGIN;
} else {
  return SIZE_LIMIT;
}

File: zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java
Comment (jiajunwang) [reply zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java:2245]: Let's just keep it strict for now.

File: zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java
Comment (jiajunwang) [reply zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java]: Then you run into the ZK server side buffer issue. The write might be done successfully, but the read will fail.
I think we agreed to keep the buffer until we bump up the ZK version to a newer one and the buffer issue should be fixed in that version.

Comment (pkuwm) [reply zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java]: @jiajunwang For default value 1000KB in ZNRecord SIZE_LIMIT, we still leave a margin that is 24 KB. But for the value that users config `jute.maxbuffer`, we should honor it as what is done in ZNRecordSerializer. We don't leave a margin in ZNRecordSerializer, right? So if user configs jute.maxbuffer, we expect them to understand it well. Say they already consider a margin in the value jute.maxbuffer, we should not(users don't expect that, either) leave another redundant margin. 

Say a user sets `jute.maxbuff` to 975KB that they think it already leaves enough room as margin, we don't want to deduct it again 975 - 24 = 951KB as the real value. It would also surprise user: if data is 960KB that user expects it to be written successfully, but it actually fails because of the extra margin.

So for the custom value if a user sets`jute.maxbuff`, I don't think we should deduct the value with the 24KB margin.

**Comment by lei-xia** [comment]
Are we still planning to merge this PR in?  What is pending here?

**Comment by pkuwm** [comment]
> Are we still planning to merge this PR in?  What is pending here?

@lei-xia Yes. It was not urgent so it was left for a while. We will still need this to make it more accurate and consistent with znrecord size limit config. Will polish the PR and get it reviewed this week.

**Review by jiajunwang — APPROVED** [review]
LGTM!
Please address the minor comments.

**Comment by pkuwm** [comment]
Thanks, @jiajunwang, for the review.

The PR is ready to be merged, approved by @jiajunwang 

If configurable write size limit for znrecord serializer is different from the default size 1 MB, especially it is set to larger than 1 MB, writing won't work because it would be blocked by this default size check in zkclient. This commit makes ZkClient to validate the data write size limit.

**Pull request merged**
